Only add camera tasks when needed

maubrunn opened the pull request:
I have noticed that we add all camera tasks in the beginning even if there is nothing subscribing to it. This slows down the frequencies of the camera topics a lot. I have added a routine that checks for subscribers to a camera topic and only then adds the task. I have also cleared up some minor bugs that I have come across when working with the driver. All is tested on our spot

EricVoll commented on spot_driver/src/spot_driver/spot_ros.py:
```diff
@@ -1310,6 +1311,26 @@ def publish_feedback(self):
         self.motion_allowed_pub.publish(self.allow_motion)
 
+    def check_for_subscriber(self):
+        # self.hand_image_mono_pub
```
```suggestion

```

EricVoll commented on spot_driver/src/spot_driver/spot_ros.py:
```diff
@@ -1310,6 +1311,26 @@ def publish_feedback(self):
+            self.hand_image_color_pub: "hand_image",
+        }
+        for pub in pubs:
+            if pub.get_num_connections() > 0 and lookup[pub] not in self.added_tasks:
```
```suggestion
            if lookup[pub] not in self.added_tasks and pub.get_num_connections() > 0:
```
I'd expect the list search to be more efficient than the get_num_connections method.

EricVoll commented on spot_driver/src/spot_driver/spot_ros.py:
```diff
@@ -1310,6 +1311,26 @@ def publish_feedback(self):
 
+    def check_for_subscriber(self):
+        # self.hand_image_mono_pub
+        pubs = [
```
The pub list and lookup dict can be defined outside of this function to avoid reinitialization at every `check_for_subscriber` call.

EricVoll commented on spot_driver/src/spot_driver/spot_wrapper.py:
```diff
@@ -1021,8 +1016,6 @@ def ensure_arm_power_and_stand(self):
             return False, "Spot with an arm is required for this service"
 
         try:
-            self._logger.info("Spot is powering on within the timeout of 20 secs")
```
Are these two lines removed on purpose?

maubrunn replied:
yes, if we try to power on spot when it is already powered on it crashes. Therefore I think it is okay to have the user power on spot manually before (or do it automatically via auto power on)

EricVoll replied:
Is there an API to check if it's powered on, and if not only then call the power on method?

heuristicus replied:
Probably use https://github.com/heuristicus/spot_ros/blob/bb82c29e9d51d12d57d53c8b4d4faecb600a1bc8/spot_driver/src/spot_driver/spot_wrapper.py#L1747-L1751

Not sure if this applies to the arm as well.

heuristicus commented on spot_driver/src/spot_driver/spot_ros.py:
```diff
@@ -108,6 +108,8 @@ def __init__(self):
         self.callbacks["side_image"] = self.SideImageCB
         self.callbacks["rear_image"] = self.RearImageCB
         self.callbacks["hand_image"] = self.HandImageCB
+        self.added_tasks = []
+        self.lookup = {}
```
These variable names aren't clear about what they're doing. I suggest something like `active_camera_tasks` and `camera_pub_to_async_task_mapping`. Long descriptive names are totally fine in my view.

heuristicus commented on spot_driver/src/spot_driver/spot_ros.py:
```diff
@@ -1426,6 +1438,13 @@ def main(self):
         )
 
+        self.lookup = {
+            self.back_image_pub: "rear_image",
```
Add front image as well?

heuristicus commented on spot_driver/src/spot_driver/spot_wrapper.py:
```diff
@@ -549,15 +549,16 @@ def __init__(
 
-            if self._robot.has_arm():
-                self._async_tasks.add_task(self._hand_image_task)
+            self.lookup = {
```
Also needs a better name. Perhaps `task_name_to_task_mapping` or something.

heuristicus commented on spot_driver/src/spot_driver/spot_wrapper.py:
```diff
@@ -1828,3 +1823,10 @@ def get_docking_state(self, **kwargs):
         return state
+
+    def update_image_tasks(self, image_name):
+        """Updates the async tasks for an image topic if there is a subscriber"""
```
I don't think this comment is actually true. From the perspective of spot_ros this might be true, but in this object subscribers don't exist. All we're doing is adding an async task based on the image name. Perhaps "Adds an async task to retrieve the images from the specified image source".

heuristicus commented on spot_driver/src/spot_driver/spot_wrapper.py:
```diff
@@ -1828,3 +1823,10 @@ def get_docking_state(self, **kwargs):
+        if self._robot.has_arm():
+            self._async_tasks.add_task(self._hand_image_task)
+
+        self._async_tasks.add_task(self.lookup[image_name])
```
Probably need guards here, check https://github.com/boston-dynamics/spot-sdk/blob/master/python/bosdyn-client/src/bosdyn/client/async_tasks.py#L26-L32 - it just appends the task to the list, not checking whether it's already there. I don't know if that is a problem, but it probably means that if this gets called multiple times the task would get executed multiple times.

heuristicus commented on spot_driver/src/spot_driver/spot_ros.py:
```diff
@@ -1426,6 +1438,13 @@ def main(self):
             "depth/frontright/depth_in_visual", Image, queue_size=10
         )
 
+        self.lookup = {
```
```suggestion
        self.camera_pub_to_async_task_mapping = {
```

heuristicus commented on spot_driver/src/spot_driver/spot_ros.py:
```diff
@@ -1310,6 +1312,16 @@ def publish_feedback(self):
         self.motion_allowed_pub.publish(self.allow_motion)
 
+    def check_for_subscriber(self):
+        for pub in list(self.lookup.keys()):
+            if (
+                self.lookup[pub] not in self.added_tasks
+                and pub.get_num_connections() > 0
+            ):
+                self.spot_wrapper.update_image_tasks(self.lookup[pub])
+                self.added_tasks.append(self.lookup[pub])
+                print("Adding :", self.lookup[pub])
```
```suggestion
        for pub in list(self.lookup.keys()):
            task_name = self.camera_pub_to_async_task_mapping[pub]
            if (
                task_name not in self.added_tasks
                and pub.get_num_connections() > 0
            ):
                self.spot_wrapper.update_image_tasks(task_name)
                self.active_camera_tasks.append(task_name)
                print(f"Detected subscriber for {task_name} task, adding task to publish")
```

Reusing the lookup a lot here, it's nicer to extract that to a variable.

maubrunn commented:
I have addressed the comments and commited the changes, however I can not test the code at the moment. I think I will be able to do it next week and will merge then.

maubrunn commented:
I have tested everything on our spot again. It is now ready to merge in my opinion

heuristicus commented:
Thanks for the contribution.
